modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -2383,6 +2424,35 @@ private CompletableFuture<Object> applyCmdWithExceptionHandling(Command cmd) {
         });
     }
 
+    private void applyCmdWithRetryOnSafeTimeReorderException(Command cmd, CompletableFuture<Object> resultFuture) {
+        raftClient.run(cmd).whenComplete((res, ex) -> {
+            if (ex != null) {
+                if (ex instanceof SafeTimeReorderException || ex.getCause() instanceof SafeTimeReorderException) {
+                    assert cmd instanceof SafeTimePropagatingCommand;
+
+                    SafeTimePropagatingCommand safeTimePropagatingCommand = (SafeTimePropagatingCommand) cmd;
+
+                    HybridTimestamp safeTimeForRetry = hybridClock.now();
+
+                    if ((cmd instanceof UpdateCommand && !((UpdateCommand) cmd).full())
```

Comment:
Why is this condition needed? Could you please add a comment explaining this?

Reply:
Also, this check seems to be fragile. What if we add an analogous request in the future, but forget to mention it here?

How about adding a marker interface common for both update commands? It might be more difficult to forget about it when adding a third 'update' command in the future.

Reply:
> Why is this condition needed? Could you please add a comment explaining this?

Added.
```
// Within primary replica it's required to update safe time in order to prevent double storage updates in case of !1PC.
// Otherwise, it may be possible that a newer entry will be overwritten by an older one that came as part of the raft
// replication flow:
// tx1 = transactions.begin();
// tx1.put(k1, v1) -> primary.apply(k1,v1) + asynchronous raft replication (k1,v1)
// tx1.put(k1, v2) -> primary.apply(k1,v2) + asynchronous raft replication (k1,v1)
// (k1,v1) replication overrides newer (k1, v2). Eventually (k1,v2) replication will restore proper value.
// However it's possible that tx1.get(k1) will see v1 instead of v2.
```

> 
> Also, this check seems to be fragile. What if we add an analogous request in the future, but forget to mention it here?
> 
> How about adding a marker interface common for both update commands? It might be more difficult to forget about it when adding a third 'update' command in the future.

I agree that given solution is fragile. It should be refactored with something more reliable in https://issues.apache.org/jira/browse/IGNITE-20124  TODO added. Here I'm talking not only about safeTime adjustment on retry but about general double updates prevention logic.

Reply:
Thanks for adding the comment, but my confusion is caused by another question: why DON'T we update safe time in case of 1PC?

Reply:
Because in case of 1PC replication goes prior to local storage update.